source/common/quic/quic_server_transport_socket_factory.cc

```diff
@@ -26,11 +27,88 @@ QuicServerTransportSocketConfigFactory::createTransportSocketFactory(
 
   auto factory = std::make_unique<QuicServerTransportSocketFactory>(
       PROTOBUF_GET_WRAPPED_OR_DEFAULT(quic_transport, enable_early_data, true),
-      context.statsScope(), std::move(server_config));
+      context.statsScope(), std::move(server_config), context.sslContextManager(), server_names);
   factory->initialize();
   return factory;
 }
 
+namespace {
+void initializeQuicCertAndKey(Ssl::TlsContext& context,
+                              const Ssl::TlsCertificateConfig& /*cert_config*/) {
+  // Convert the certificate chain loaded into the context into PEM, as that is what the QUICHE
+  // API expects. By using the version already loaded, instead of loading it from the source,
+  // we can reuse all the code that loads from different formats, allows using passwords on the key,
+  // etc.
+  std::vector<std::string> chain;
+  auto process_one_cert = [&](X509* cert) {
+    bssl::UniquePtr<BIO> bio(BIO_new(BIO_s_mem()));
+    int result = PEM_write_bio_X509(bio.get(), cert);
+    ASSERT(result == 1);
+    BUF_MEM* buf_mem = nullptr;
+    result = BIO_get_mem_ptr(bio.get(), &buf_mem);
+    std::string cert_str(buf_mem->data, buf_mem->length);
+    std::istringstream pem_stream(cert_str);
+    auto pem_result = quic::ReadNextPemMessage(&pem_stream);
+    RELEASE_ASSERT(pem_result.status == quic::PemReadResult::Status::kOk,
```

Comment:
This is run anytime the certificates are updated, including during SDS. The reason I left it as a release-assert is that:

1) The data has been loaded by boringssl in the tls transport socket already, and then boringssl wrote it out as PEM. So we know it's a valid PEM certificate message; an error would have been generated earlier if it were somehow invalid
2) This shouldn't be possible to fail, but if it did, it could have security implications. But because it cannot fail, I don't want to add an error case here because I can't test it; it would be unreachable.

Reply:
It shouldn't fail today but quic::ReadNextPemMessage() might be changed in the future, i.e. adding more validations to the certs. IMO it's worth to throw an error here even with uncovered code.

Reply:
Good point about future changes to how that function works. I'll make this change.